samples/boards/nrf/s2ram/src/main.c

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2022 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/zephyr.h>
+
+#include <zephyr/pm/pm.h>
+#include <zephyr/pm/policy.h>
+
+#include <zephyr/arch/common/pm_s2ram.h>
+
+#include <hal/nrf_power.h>
+#include <hal/nrf_regulators.h>
+#include <hal/nrf_gpio.h>
+
+static void system_off(void)
+{
+	nrf_regulators_system_off(NRF_REGULATORS);
+}
+
+static void do_suspend(void)
+{
+	irq_lock();
+
+	/*
+	 * Save the CPU context (including the return address),set the SRAM
+	 * marker and power off the system.
+	 */
+	arch_pm_s2ram_suspend(system_off);
+
+	/*
+	 * XXX: On resuming we return exactly *HERE*
+	 */
+}
+
+void main(void)
+{
+	printk("S2RAM demo (%s)\n", CONFIG_BOARD);
+
+	/*
+	 * Enable wake-up button and LED1
+	 */
+	nrf_gpio_cfg_input(23,  NRF_GPIO_PIN_PULLUP);
+	nrf_gpio_cfg_sense_set(23, NRF_GPIO_PIN_SENSE_LOW);
+	nrf_gpio_cfg_output(28);
+
+	/*
+	 * Sleep enough time to trigger S2RAM
+	 */
+	while (true) {
+		k_msleep(1000);
+	}
+}
+
+void pm_state_set(enum pm_state state, uint8_t substate_id)
+{
+	if (state != PM_STATE_SUSPEND_TO_RAM) {
+		return;
+	}
+
+	do_suspend();
+}
+
+void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
+{
+	if (state != PM_STATE_SUSPEND_TO_RAM) {
+		return;
+	}
+
+	/*
+	 * Turn the LED3 on
+	 */
+	nrf_gpio_cfg_output(30);
+
+	/*
+	 * XXX: Here we should restore all the peripherals of the platforms
+	 * that will not be restored by the PM subsystem: NVIC, PMU, RTC, etc...
+	 */
+
+	irq_unlock(0);
+
+	while (true) {
+		/* stop here */
+	}
```

Comment:
I assume this is here because this sample is not 100% functional, so it should probably be made clear in the sample readme.

Reply:
IMO the sample should be required to at least put the system back into a state where you can use `k_sleep` etc.
If that is too hard to do in a dedicated sample, what hope does a real application have?

Reply:
@JordanYates 

This PR is about adding the infrastructure for S2RAM (mainly to discuss about the abstraction and the hooks). The sample is there to show how (in thory) the abstraction should be used. I could leave that out from this PR if that's ok but I think it's still better to have "something" at least.

The problem about the non-functional test is that we are missing a bunch of other pieces to have the test fully functional, for example a way to restore the NVIC and PMU states without relying on the device PM and that's definitely outside the scope of this PR.

Reply:
If that's the case the `XXX` comment should clearly spell that out. I would read that comment and then spend who knows how long trying to find the API's etc to put everything back together.

Reply:
@gmarull @JordanYates I update both the README and the XXX part. Let me know if that's ok with you.

Reply:
Much clearer now, thanks